includes/class-omise-payment-result.php

```diff
@@ -0,0 +1,129 @@
+	 * @var \OmiseCharge
+	 */
+	protected $charge;
+
+	/**
+	 * @var string
+	 */
+	protected $context;
+
+	/**
+	 * @param \WC_Abstract_Order $order
+	 * @param \OmiseCharge       $charge
+	 * @param string             $context
+	 */
+	public function __construct( $order, $charge, $context ) {
+		$this->order   = $order;
+		$this->charge  = $charge;
+		$this->context = $context;
+	}
+
+	/**
+	 * @param string       $order_id
+	 * @param string       $charge
+	 * @param string|null  $context
+	 *
+	 * @static
+	 */
+	public static function callback( $order_id, $charge, $context = null ) {
+		$result = new self( wc_get_order( $order_id ), unserialize( $charge ), $context );
+		$result->resolve();
+	}
+
+	/**
+	 * Resolving order status based on
+	 * the result of a particulalr payment.
+	 */
+	public function resolve() {
+		switch ( strtolower( $this->charge['status'] ) ) {
+			case 'successful':
+				$this->payment_successful();
+				break;
+
+			case 'failed':
+				$this->payment_failed();
+				break;
+
+			case 'pending':
+				if ( ! $this->charge['capture'] && $this->charge['authorized'] ) {
+					$this->payment_authorized();
+				} else {
+					$this->payment_pending();
+				}
+				break;
+
+			default:
+				$this->order->add_order_note( __( 'Unrecognized Omise Charge status.', 'omise' ) );
+				break;
+		}
+	}
+
+	/**
+	 * Resolving a case of charge status: successful.
+	 */
+	public function payment_successful() {
+		if ( $this->order->has_status( 'processing' ) ) {
+			return;
+		}
+
+		$message = wp_kses( __( 'OMISE: Payment successful.<br/>An amount of %1$s %2$s has been paid', 'omise' ), array( 'br' => array() ) );
+		$this->order->add_order_note( sprintf( $message, $this->order->get_total(), $this->order->get_currency() ) );
+		$this->order->payment_complete();
+	}
+
+	/**
+	 * Resolving a case of charge status: failed.
+	 */
+	public function payment_failed() {
+		if ( $this->order->has_status( 'failed' ) ) {
+			return;
+		}
+
+		$failure_message = $this->charge['failure_message'] . ' (code: ' . $this->charge['failure_code'] . ')';
+		$this->order->add_order_note( sprintf( wp_kses( __( 'OMISE: Payment failed.<br/>%s', 'omise' ), array( 'br' => array() ) ), $failure_message ) );
+		$this->order->update_status( 'failed' );
+	}
+
+	/**
+	 * Resolving a case of charge status: pending.
+	 */
+	public function payment_pending() {
+		$message = wp_kses( __(
+			'Omise: The payment is being processed.<br/>
+			Depending on the payment provider, this may take some time to process.<br/>
+			Please do a manual \'Sync Payment Status\' action from the <strong>Order Actions</strong> panel, or check the payment status directly at the Omise Dashboard later.',
+			'omise'
+		), array( 'br' => array(), 'strong' => array() ) ) ;
+
+		$this->order->add_order_note( $message );
+		$this->order->update_status( 'on-hold' );
+	}
+
+	/**
+	 * Resolving a case of an authorized charge.
+	 */
+	public function payment_authorized() {
+		$message = wp_kses( __(
```

Comment:
If manual capture is selected, this may cause adding duplicate note to order for charge authorisation. Similar note in `Omise_Payment_Creditcard`.

Reply:
@mayurkathale Can I have more detail about this case? 
Here is the manual-capture case.
<img width="1281" alt="Screen Shot 2563-07-15 at 16 16 29" src="https://user-images.githubusercontent.com/2154669/87527597-8e2eca00-c6b6-11ea-8a68-d5ad720c3c1a.png">

Reply:
@guzzilar maybe need to do some activity on the website which might help to worker jobs to get executed. This should be done before payment capture.

Reply:
I understand the issue now.
Just pushed a new commit here: https://github.com/omise/omise-woocommerce/pull/177/commits/84d09ab068eb94e0103514d5293b608d60390f6f.
Thanks for catching it 👍👍